internal/oci/container.go

```diff
@@ -344,6 +386,109 @@ func (c *Container) exitFilePath() string {
 	return filepath.Join(c.dir, "exit")
 }
 
+// IsRunning is a function that checks if a specific pid is running
+// it is used to check a container state when we don't want (or don't trust) a `$runtime state` call
+func (c *Container) IsRunning() bool {
+	_, err := c.pid()
+	if err != nil {
+		logrus.Errorf("checking if pid of %s is running failed: %v", c.id, err)
+		return false
+	}
+
+	return true
+}
+
+// Pid is the external, locking function that returns the container's
+// pid, if it's running, and actually the container process
+func (c *Container) Pid() (int, error) {
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+	return c.pid()
+}
+
+// pid is the internal, non-locking version of pid. It checks
+// that we have a Pid defined in the state, that pid can be found
+// and it is the same process that was originally started by the runtime
+func (c *Container) pid() (int, error) {
+	if c.state == nil {
+		return -1, errors.New("state not initialized")
+	}
+	pid := c.state.InitialPid
+	if pid <= 0 {
+		return -1, errors.New("pid not initialized")
+	}
+
+	// container has stopped (as pid is initialized but the runc state has overwritten it)
+	if c.state.Pid == 0 {
+		return 0, findprocess.ErrNotFound
+	}
+
+	if _, err := findprocess.FindProcess(pid); err != nil {
+		return -1, err
+	}
+
+	if err := c.verifyPid(pid); err != nil {
+		return -1, err
+	}
+	return pid, nil
+}
+
+// findAndReleasePid attempts to find the container's pid
+// and if it's found, release it. In between these steps,
+// it also verified the pid running is the same as was originally
+// started by the runtime
+func (c *Container) findAndReleasePid() (bool, error) {
+	pid := c.state.InitialPid
+	found := false
+
+	process, err := findprocess.FindProcess(pid)
+	if err == findprocess.ErrNotFound {
+		return found, nil
+	}
+	if err != nil {
+		return found, err
+	}
+
+	if err := c.verifyPid(pid); err != nil {
+		return found, err
+	}
+
+	found = true
+	if err := process.Release(); err != nil {
```

Comment:
I'd still try to release it anyway.  If it is not a direct child an error will be raised, and if it is a direct child then we are sure to free the process in case the verification step fails

Reply:
good idea, fixed as suggested